mix/QBigInt.h

```diff
@@ -0,0 +1,97 @@
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/** @file QBigInt.h
+ * @author Yann yann@ethdev.com
+ * @date 2015
+ * Represent a big integer (u256, bigint) to be used in QML.
+ */
+
+#pragma once
+
+#include "boost/variant.hpp"
+#include "boost/variant/multivisitors.hpp"
+#include <QObject>
+#include <libdevcore/CommonJS.h>
+#include <libdevcore/Common.h>
+
+using namespace dev;
+
+namespace dev
+{
+namespace mix
+{
+
+using BigIntVariant = boost::variant<dev::u256, dev::bigint>;
+
+struct add: public boost::static_visitor<BigIntVariant>
+{
+	template<class T1, class T2>
+	BigIntVariant operator()(T1 const& _value, T2 const& _otherValue) const { return _value + _otherValue; }
+};
+
+struct subtract: public boost::static_visitor<BigIntVariant>
+{
+	template<class T1, class T2>
+	BigIntVariant operator()(T1 const& _value, T2 const& _otherValue) const { return _value - _otherValue; }
+};
+
+struct multiply: public boost::static_visitor<BigIntVariant>
+{
+	template<class T1, class T2>
+	BigIntVariant operator()(T1 const& _value, T2 const& _otherValue) const { return _value * _otherValue; }
+};
+
+struct divide: public boost::static_visitor<BigIntVariant>
+{
+	template<class T1, class T2>
+	BigIntVariant operator()(T1 const& _value, T2 const& _otherValue) const { return _value / _otherValue; }
+};
+
+class QBigInt: public QObject
+{
+	Q_OBJECT
+
+public:
+	QBigInt(QObject* _parent = 0): QObject(_parent), m_internalValue(dev::u256(0)) {}
+	QBigInt(dev::u256 const& _value, QObject* _parent = 0): QObject(_parent), m_internalValue(_value) {}
+	QBigInt(dev::bigint const& _value, QObject* _parent = 0): QObject(_parent), m_internalValue(_value) {}
+	QBigInt(BigIntVariant const& _value, QObject* _parent = 0): QObject(_parent), m_internalValue(_value){}
+	~QBigInt() {}
+
+	/// @returns the current used big integer.
+	BigIntVariant internalValue() { return m_internalValue; }
+	/// @returns a string representation of the big integer used. Invokable from QML.
+	Q_INVOKABLE QString value() const;
+	/// Set the value of the BigInteger used. Will use u256 type. Invokable from QML.
+	Q_INVOKABLE void setValue(QString const& _value) { m_internalValue = dev::jsToU256(_value.toStdString()); }
+	/// Subtract by @a _value. Invokable from QML.
+	Q_INVOKABLE QBigInt* subtract(QBigInt* const& _value) const;
+	/// Add @a _value to the current big integer. Invokable from QML.
+	Q_INVOKABLE QBigInt* add(QBigInt* const& _value) const;
+	/// Multiply by @a _value. Invokable from QML.
+	Q_INVOKABLE QBigInt* multiply(QBigInt* const& _value) const;
+	/// divide by @a _value. Invokable from QML.
+	Q_INVOKABLE QBigInt* divide(QBigInt* const& _value) const;
+
+protected:
+	 BigIntVariant m_internalValue;
```

Comment:
Yes I tried but we cannot use Template with a QObject class. And with
Template, we have to know the type to use (bigint or u256) at compile time.
With boost::variant. We just have to set a pool of type to manage and the
rest is done during the runtime.

2015-01-19 16:48 GMT+01:00 Arkadiy Paronyan <notifications@github.com>:

> In mix/QBigInt.h
> <https://github.com/ethereum/cpp-ethereum/pull/819#discussion-diff-23167311>
> :
>
> > +	BigIntVariant internalValue() { return m_internalValue; }
> > +	/// @returns a string representation of the big integer used. Invokable from QML.
> > +	Q_INVOKABLE QString value() const;
> > +	/// Set the value of the BigInteger used. Will use u256 type. Invokable from QML.
> > +	Q_INVOKABLE void setValue(QString const& _value) { m_internalValue = dev::jsToU256(_value.toStdString()); }
> > +	/// Subtract by @a _value. Invokable from QML.
> > +	Q_INVOKABLE QBigInt* subtract(QBigInt* const& _value) const;
> > +	/// Add @a _value to the current big integer. Invokable from QML.
> > +	Q_INVOKABLE QBigInt* add(QBigInt* const& _value) const;
> > +	/// Multiply by @a _value. Invokable from QML.
> > +	Q_INVOKABLE QBigInt* multiply(QBigInt* const& _value) const;
> > +	/// divide by @a _value. Invokable from QML.
> > +	Q_INVOKABLE QBigInt* divide(QBigInt* const& _value) const;
> > +
> > +protected:
> > +	 BigIntVariant m_internalValue;
>
> Using variant here makes code quite complex. Maybe use template base class
> instead?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ethereum/cpp-ethereum/pull/819/files#r23167311>.
>